coalib/core/Bear.py

```diff
@@ -0,0 +1,469 @@
+        >>> file_dict = {'file1.txt': ['']}
+        >>> bear = Bear(section, file_dict)
+        >>> bear.dependency_results
+        {}
+        >>> dependency_bear = Bear(section, file_dict)
+        >>> bear.add_dependency_results(dependency_bear, [1, 2, 3])
+        >>> bear.dependency_results
+        {<class 'coalib.core.Bear.Bear'>: [1, 2, 3]}
+
+        :param bear:
+            The bear instance to add results for.
+        :param dependency_results:
+            The results to add.
+        """
+        bear_type = type(bear)
+
+        if bear_type not in self._dependency_results:
+            current_results = []
+            self._dependency_results[bear_type] = current_results
+        else:
+            current_results = self._dependency_results[bear_type]
+
+        current_results += dependency_results
+
+    @property
+    def dependency_results(self):
+        """
+        Contains all dependency results.
+
+        This variable gets set during bear execution from the core and can be
+        used from ``analyze``.
+
+        Modifications to the returned dictionary lead to undefined behaviour.
+
+        >>> section = Section('my-section')
+        >>> file_dict = {'file1.txt': ['']}
+        >>> bear = Bear(section, file_dict)
+        >>> bear.dependency_results
+        {}
+        >>> dependency_bear = Bear(section, file_dict)
+        >>> bear.add_dependency_results(dependency_bear, [1, 2, 3])
+        >>> bear.dependency_results
+        {<class 'coalib.core.Bear.Bear'>: [1, 2, 3]}
+
+        :return:
+            A dictionary with bear-types as keys and their results received.
+        """
+        return self._dependency_results
+
+    @classmethod
+    def get_metadata(cls):
+        """
+        :return:
+            Metadata for the ``analyze`` function extracted from its signature.
+            Excludes parameter ``self``.
+        """
+        return FunctionMetadata.from_function(
+            cls.analyze,
+            omit={'self'})
+
+    # FIXME Make this a @classproperty.
+    @classmethod
+    def get_non_optional_settings(cls):
+        """
+        This method has to determine which settings are needed by this bear.
+        The user will be prompted for needed settings that are not available
+        in the settings file so don't include settings where a default value
+        would do.
+
+        :return: A dictionary of needed settings as keys and a tuple of help
+                 text and annotation as values
+        """
+        return cls.get_metadata().non_optional_params
+
+    @classmethod
+    def __json__(cls):
+        """
+        Override JSON export of ``Bear`` class.
+        """
+        # Those members get duplicated if they aren't excluded because they
+        # exist also as fields.
+        excluded_members = {'can_detect', 'maintainers', 'maintainers_emails'}
+
+        # json cannot serialize properties, so drop them
+        data = {
+            key: value
```

Comment:
we can't^^ We need to implement `str` on those which don't have one / we want to change so they look better^^